### PR TITLE
Distances and zones of replics

### LINKED ISSUE #201: [4pt] Replicas weight configuration
Currently I found no way to configure [weights and zones](https://www.tarantool.io/en/doc/2.2/reference/reference_rock/vshard/vshard_admin/#replica-weights) for replicas and routers.

I think it would be nice to have such a feature in the UI.

### PR DESCRIPTION
**Pull request by Steap2448**
New section `zone_distances` in clusterwide config and `zone` option in `servers` section are implemented.

GraphQL is updated accordingly:

```graphql
query { servers {zone} }
```

```graphql
mutation {
    cluster {
        edit_topology(servers: [{uuid: "..." zone: "..."}]){}
    }
}
```

I didn't forget about

- [x] Tests
- [x] Changelog
- [x] Documentation (ldoc)

Close #201.

### CODE REVIEW

File: cartridge/lua-api/edit-topology.lua [hunk 1]
```diff
@@ -22,6 +22,7 @@ local topology_cfg_checker = {
     failover = '?',
     servers = 'table',
     replicasets = 'table',
+    zone_distances = '?table',
```
Comment (rosik) [inline cartridge/lua-api/edit-topology.lua:25]: Redundant

File: cartridge/lua-api/edit-topology.lua [hunk 2]
```diff
@@ -261,6 +271,7 @@ local function edit_topology(args)
     checks({
         replicasets = '?table',
         servers = '?table',
+        zone_distances = '?table'
```
Comment (rosik) [inline cartridge/lua-api/edit-topology.lua:274]: Redundant

File: cartridge/lua-api/edit-topology.lua [hunk 3]
```diff
@@ -337,14 +348,21 @@ local function edit_topology(args)
...
     end
 
-    local ok, err = twophase.patch_clusterwide({topology = topology_cfg})
+    local zone_distances = table.deepcopy(args.zone_distances) or {}
+    topology.validate_distances(zone_distances)
+
+    local ok, err = twophase.patch_clusterwide({
+        topology = topology_cfg,
+        zone_distances = zone_distances
+    })
```
Comment (rosik) [inline cartridge/lua-api/edit-topology.lua:357]: Redundant

File: cartridge/lua-api/edit-topology.lua [hunk 4]
```diff
@@ -337,14 +348,21 @@ local function edit_topology(args)
...
     if not ok then
         return nil, err
     end
 
     local ret = {
         replicasets = {},
         servers = {},
+        zone_distances = {},
```
Comment (rosik) [inline cartridge/lua-api/edit-topology.lua:365]: Redundant

File: cartridge/lua-api/edit-topology.lua [hunk 5]
```diff
@@ -363,6 +381,8 @@ local function edit_topology(args)
         table.insert(ret.replicasets, topology.replicasets[rpl.uuid])
     end
 
+    ret.zone_distances = table.deepcopy(zone_distances)
```
Comment (rosik) [inline cartridge/lua-api/edit-topology.lua:384]: Redundant

File: cartridge/lua-api/get-topology.lua [hunk 6]
```diff
@@ -219,9 +219,12 @@ local function get_topology()
         end
     end
 
+    local zone_distances = confapplier.get_readonly('zone_distances')
+
     return {
         servers = servers,
         replicasets = replicasets,
+        zone_distances = zone_distances,
```
Comment (rosik) [inline cartridge/lua-api/get-topology.lua:227]: Redundant

File: cartridge/roles/vshard-storage.lua [hunk 7]
```diff
@@ -14,6 +14,12 @@ local function apply_config(conf, _)
     local my_replicaset = conf.topology.replicasets[box.info.cluster.uuid]
     local group_name = my_replicaset.vshard_group or 'default'
     local vshard_cfg = vshard_utils.get_vshard_config(group_name, conf)
+    if vshard_cfg.weights ~= nil then
+        vshard_cfg.weights = nil
+    end
+    if vshard_cfg.zone ~= nil then
```
Comment (rosik) [inline cartridge/roles/vshard-storage.lua:20]: Is conditional expression necessary?

File: cartridge/topology.lua [hunk 8]
```diff
@@ -56,6 +57,19 @@ vars:new('known_roles', {
...
+            ['3'] = 2,
+            ['4'] = 3,
+        },
+        ['2'] = {
+            ['1'] = 10,
+            ['2'] = 0,
+            ['3'] = 10,
+            ['4'] = 20,
+        },
+    }
```
Comment (rosik) [inline cartridge/topology.lua:72]: Redundant?

File: cartridge/vshard-utils.lua [hunk 9]
```diff
@@ -266,6 +266,8 @@ local function validate_config(conf_new, conf_old)
                 validate_group_upgrade(name, topology_new, topology_old)
             end
         end
+
+        topology.validate_distances(conf_new.zone_distances);
```
Comment (rosik) [inline cartridge/vshard-utils.lua:270]: I suggest moving it's implementation to this module (similar to `validate_group_weights`)

**Review by rosik — CHANGES_REQUESTED** [review]
I don't like the idea of extending `edit_topology`/`get_topology` with `zone_distances`. At first, `zone_distances.yml` is a separate config section, it's not a part of `topology.yml`. And secondly, Lua API is similar to the GraphQL, but `zone_distances` can't be added there (due to graphql types strictness).

File: test/integration/vshard_distances_test.lua [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    helpers.retrying({}, function()
+        local res = get_config()
+        t.assert_items_equals(res.zone_distances, distances)
+        t.assert_equals(res.topology.servers[M1.instance_uuid].zone, '1')
+        t.assert_equals(res.topology.servers[R1.instance_uuid].zone, '2')
+        t.assert_equals(res.topology.servers[R2.instance_uuid].zone, '3')
+    end)
+
+    local response = M1.net_box:eval(q_get_priority, {helpers.uuid('a')})
+    check_order(response,
```
[the remark this answers is not in the document]
Comment (Steap2448) [reply test/integration/vshard_distances_test.lua:172]: It doesn't take order into account

[the remark this answers is not in the document]
Comment (rosik) [reply test/integration/vshard_distances_test.lua:172]: Then `assert_equals`

File: cartridge/vshard-utils.lua [hunk 11]
```diff
@@ -387,6 +436,26 @@ local function get_vshard_config(group_name, conf)
...
+        zone = nil
+    end
+
+    local zone_distances = confapplier.get_deepcopy('zone_distances')
+    if zone_distances == nil then
+        zone_distances = nil
+    else
+        for zone1, zone in pairs(zone_distances) do
+            for zone2, distance in pairs(zone) do
+                if distance == nil then
```
Comment (olegrok) [inline cartridge/vshard-utils.lua:452]: What's about `distance == 0`?

**Review by rosik — APPROVED** [review]

File: test/integration/zones_test.lua [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+        "bad argument params.zone to __edit_server"..
+        " (?string expected, got number)",
+        set_zones, {[g.A2.instance_uuid] = 2}
+    )
+end
+
+function g.test_distances()
+    local q_get_priority = [[
+        local replicaset = require('vshard').router.routeall()[...]
+        require('log').info(replicaset.priority_list)
```
Comment (olegrok) [inline test/integration/zones_test.lua:121]: Remove if it's a debug. Add more verbose message if it's really needed.

**Review by olegrok — COMMENTED** [review]
Thanks for your patch. Consider my minor comments below.

File: cartridge/vshard-utils.lua (hunk 11 above)
Comment (rosik) [reply cartridge/vshard-utils.lua:452]: 0 is ok for vshard.

File: cartridge/vshard-utils.lua [hunk 13]
```diff
@@ -56,6 +56,53 @@ local function validate_group_weights(group_name, topology)
...
+
+    for zone1, v in pairs(zone_distances) do
+        ValidateConfigError:assert(
+            type(zone1) == 'string',
+            'Zone label must be a string, got %s', type(zone1)
+        )
+
+        ValidateConfigError:assert(
+            type(v) == 'table',
+            'Zone %s must be map of relative weights of other zones, got %s',
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/vshard-utils.lua]: Agree

File: cartridge/vshard-utils.lua [hunk 14]
```diff
@@ -395,6 +464,8 @@ local function get_vshard_config(group_name, conf)
         collect_bucket_garbage_interval = vshard_groups[group_name].collect_bucket_garbage_interval,
         rebalancer_disbalance_threshold = vshard_groups[group_name].rebalancer_disbalance_threshold,
         sharding = sharding,
+        weights = zone_distances,
```
[the remark this answers is not in the document]
Comment (rosik) [reply cartridge/vshard-utils.lua:472]: Fixed

File: test/integration/zones_test.lua (hunk 12 above)
Comment (rosik) [reply test/integration/zones_test.lua]: Fixed

**Review by olegrok — APPROVED** [review]

**Pull request merged**